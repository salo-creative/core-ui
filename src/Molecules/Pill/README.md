# Pill

The Pill component is meant to display values as tags. If given an onRemove prop, it will render a close button as well. Clicking the close button fires the onRemove event specified. It can also take and display a loading prop.

## Usage

Install

```javascript
yarn add @salo/core-ui
```

Include

```javascript
import Pill from '@salo/core-ui/Molecules/Pill';
```

Implement as follows

```javascript
<Pill
  value="Pill"
  onRemove={ (value) => remove(value) }
  loading={loading}
/>
```