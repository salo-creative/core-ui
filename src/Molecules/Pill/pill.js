import React from 'react';
import PropTypes from 'prop-types';
import Icon from '@salo/icons';
import Loader from '../Loader';

// COMPONENTS & STYLES
import { PillWrapper, HiddenCloseButton } from './pills.styles';

const Pill = (props) => {
  const {
    background,
    color,
    height,
    label,
    loading,
    margin,
    onRemove,
    padding,
    value
  } = props;
 
  return (
    <PillWrapper
      color={ color }
      background={ background }
      height={ height }
      margin={ margin }
      loading={ loading }
      padding={ padding }
    >
      { loading && (
        <Loader
          display={ true }
          loaderProps={ { size: 40, position: 'absolute' } }
          appearance='light'
        />
      ) }
      { label }
      { onRemove && (
      <HiddenCloseButton
        onClick={ () => onRemove(value) }
        disabled={ loading }
        loading={ loading }
        padding={ padding }
      >
        <Icon
          size={ 20 }
          icon='close'
        />
      </HiddenCloseButton>
      ) }
      
    </PillWrapper>
  );
};

Pill.defaultProps = {
  background: 'grey',
  color: 'black',
  height: '4rem',
  label: '',
  loading: false,
  margin: '',
  onRemove: null,
  padding: '1.5rem',
  value: ''
};

Pill.propTypes = {
  background: PropTypes.string,
  color: PropTypes.string,
  height: PropTypes.string,
  label: PropTypes.string,
  loading: PropTypes.bool,
  margin: PropTypes.string,
  onRemove: PropTypes.func,
  padding: PropTypes.string,
  value: PropTypes.string
};

export default Pill;