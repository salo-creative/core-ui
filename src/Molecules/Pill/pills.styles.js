import styled, { css, keyframes } from 'styled-components';

// HELPERS
import { boxShadow } from '../../helpers/colours';

const rotate = keyframes`
  from {
    transform: rotate(0deg);
  }

  to {
    transform: rotate(360deg);
  }
`;

export const PillWrapper = styled.div`
    background-color: ${ ({ theme, background }) => (theme[background] ? theme[background] : theme.primary) };
    border: ${ ({ border }) => border };
    border-radius: 5rem;
    box-shadow: ${ ({ shadow }) => shadow || boxShadow() };
    color: ${ ({ theme, color, loading, background, inlineLoader }) => (!inlineLoader && loading ? theme[background] : theme[color]) };
    display: inline-flex;
    font-size: ${ ({ fontSize }) => fontSize };
    margin: ${ ({ margin }) => (margin || '0 0.5rem 0 0') };
    padding: ${ ({ padding }) => padding };
    position: relative;
    transition: all 0.3s linear;

    path {
        fill: ${ ({ loading, color, background, theme, inlineLoader }) => (!inlineLoader && loading ? theme[background] : theme[color]) };
        transition: fill 0.3s linear;
    }
`;

export const HiddenButton = styled.button`
    background-color: transparent;
    border: 0;
    cursor: pointer;
    display: inline-flex;
    margin: 0;
    min-width: 2rem;
    opacity: 0.6;
    padding: 0;
    transition: opacity 0.3s linear;
    visibility: ${ ({ isLoading, inlineLoader }) => (!inlineLoader && isLoading ? 'hidden' : 'visible') };

    &:hover {
        opacity: 1;
    }

    svg {
        transform-origin: center center;
        ${ ({ isLoading }) => (isLoading ? css`animation: ${ rotate } 2s linear infinite;` : '') }
    }
`;

export const Text = styled.span`
  margin-right: 1.5rem;
`;