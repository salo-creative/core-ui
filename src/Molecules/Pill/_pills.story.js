import React from 'react';
import {
  withKnobs, select, text, boolean
} from '@storybook/addon-knobs';
import { RenderWithProps, Store } from '@jamesbliss/storybook-state';

// load tests
import { withTests } from '@storybook/addon-jest';
import results from '../../../.storybook/jest-test-results.json';

// FEATURED COMPONENT
import Pill from './pill';

// README //
import README from './README.md';

// Start of story logic
export const SinglePill = () => {
  // STORE
  const store = new Store({});
  // KNOBS
  const value = text('Value', '1');
  const label = text('Label', 'Single Pill');
  const color = select(
    'Color',
    ['blue', 'red', 'orange', 'navy', 'green', 'darkGrey', 'yellow', 'white', 'black'],
    'black'
  );
  const background = select(
    'Background',
    ['blue', 'red', 'orange', 'navy', 'green', 'darkGrey', 'yellow', 'white', 'grey'],
    'grey'
  );
  const loading = boolean('Show loading state', false);
  return (
    <RenderWithProps store={ store }>
      <Pill
        value={ value }
        label={ label }
        color={ color }
        background={ background }
        onRemove={ id => console.log(id) }
        loading={ loading }
      />
    </RenderWithProps>
  );
};

SinglePill.story = {
  decorators: [
    withKnobs,
    withTests({
      results
    })
  ],
  parameters: {
    jest: ['pills'],
    info: {
      propTablesExclude: [RenderWithProps]
    },
    notes: README
  }
};

export default {
  title: 'Molecules/Pill'
};