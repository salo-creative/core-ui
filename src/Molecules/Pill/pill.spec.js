import React from 'react';
import 'jest-styled-components';

import { renderWithTheme } from '../../../test';

// COMPONENT
import Pill from './index';

test('Renders against snapshot', async () => {
  const { container } = renderWithTheme(
    <Pill />
  );
  expect(container.firstChild).toMatchSnapshot();
});