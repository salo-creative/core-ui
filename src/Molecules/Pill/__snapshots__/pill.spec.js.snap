// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders against snapshot 1`] = `
.c0 {
  background: #e9ecef;
  border: none;
  border-radius: 5rem;
  box-shadow: 2px 2px 2px 0px rgba(0,0,0,0.2);
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  font-size: 1.6rem;
  margin: 0 0.5rem 0 0;
  padding: 1.5rem;
  position: relative;
  -webkit-transition: all 0.3s linear;
  transition: all 0.3s linear;
  outline: none;
}

.c0:focus-within {
  border-color: #00ACED;
}

.c0 path {
  -webkit-transition: fill 0.3s linear;
  transition: fill 0.3s linear;
}

.c1 {
  margin: 0;
}

<div
  class="c0"
  color="black"
  font-size="1.6rem"
  height="4rem"
>
  <span
    class="c1"
  />
</div>
`;
